Rental & Delegation NFT - ERC721 Extension (#5501)
ethereum/EIPs: EIPS/eip-5501.md

---
eip: 5501
title: Rental & Delegation NFT - EIP-721 Extension
description: Adds a conditional time-limited user role to EIP-721. This role can be delegated or borrowed.
author: Jan Smrža (@smrza), David Rábel (@rabeles11), Tomáš Janča <tomas.janca@jtbstorage.eu>, Jan Bureš (@JohnyX89), DOBBYLABS (@DOBBYLABS)
discussions-to: https://ethereum-magicians.org/t/eip-tbd-rental-delegation-nft-erc-721-extension/10441
status: Draft
type: Standards Track
category: ERC
created: 2022-08-18

> requires: 165, 721

Reviewer: ```suggestion
requires: 165, 721, 4400, 4907
```
Required: add all EIPs referenced to `requires` header.

---

## Abstract
The following standard proposes an additional `user` role for [EIP-721](./eip-721.md). This role grants the permission to use the NFT with no ability to transfer or set users. It has an expiry and a flag if the token is borrowed or not. `Owner` can delegate the NFT for usage to hot wallets or lend the NFT. If the token is borrowed, not even the owner can change the user until the status expires or both parties agree to terminate. This way, it is possible to keep both roles active at the same time.

## Motivation
Collectibles, gaming assets, metaverse, event tickets, music, video, domains, real item representation are several among many NFT use cases. With [EIP-721](./eip-721.md) only the owner can reap the benefits. However, with most of the utilities it would be beneficial to distinguish between the token owner and its user. For instance music or movies could be rented. Metaverse lands could be delegated for usage. 

> The two reasons why to set the user are:

Reviewer: ```suggestion
The two reasons why to set the user are:

```
Required: formatting.

* **delegation** - Assign user to your hot wallet to interact with applications securely. In this case, the owner can change the user at any time.
* **renting** - This use case comes with additional requirements. It is needed to terminate the loan once the established lending period is over. This is provided by `expires` of the user. It is also necessary to protect the borrower against resetting their status by the owner. Thus, `isBorrowed` check must be implemented to disable the option to set the user before the contract expires.

> The most common use cases for having an additional user role are:

Reviewer: ```suggestion
The most common use cases for having an additional user role are:

```
Required: formatting.

* **delegation** - For security reasons.
* **gaming** - Would you like to try a game (or particular gaming assets) but are you unsure whether or not you will like it? Rent assets first.
* **guilds** - Keep the owner of the NFTs as the multisig wallet and set the user to a hot wallet with shared private keys among your guild members.
* **events** - Distinguish between `ownerOf` and `userOf`. Each role has a different access.
* **social** - Differentiate between roles for different rooms. For example owner has read + write access while userOf has read access only.

> This proposal is a follow up on [EIP-4400](./eip-4400.md) and [EIP-4907](./eip-4907.md) and introduces additional upgrades for lending and borrowing which include:

Reviewer: ```suggestion
This proposal is a follow up on [EIP-4400](./eip-4400.md) and [EIP-4907](./eip-4907.md) and introduces additional upgrades for lending and borrowing which include:

```
Required: formatting.

Building the standard with additional isBorrowed check now allows to create rental marketplaces which can set the user of NFT without the necessary staking mechanism. With current standards if a token is not staked during the rental period, the owner can simply terminate the loan by setting the user repeatedly. This is taken care of by disabling the function if the token is borrowed which in turn is providing the owner additional benefits. They can keep the token tied to their wallet, meaning they can still receive airdrops, claim free mints based on token ownership or otherwise use the NFT provided by third-party services for owners. They can also keep the NFT listed for sale. Receiving airdrops or free mints was previously possible but the owner was completely reliant on the implementation of rental marketplaces and their discretion.

Decentralized applications can now differentiate between ownerOf and userOf while both statuses can coexist.
  
## Specification
The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

> **Every compliant contract MUST implement the `IERC5501` interface. This extension is OPTIONAL for [EIP-721](./eip-721.md) contracts.**

Reviewer: ```suggestion
**Every compliant contract MUST implement the `IERC5501` interface. This extension is OPTIONAL for [EIP-721](./eip-721.md) contracts.**

```
Required: formatting.

```solidity
/**
 * @title IERC5501: Rental & Delegation NFT - EIP-721 Extension

>  * @dev See https://eips.ethereum.org/EIPS/eip-5501

Reviewer: ```suggestion
```
Required: no external links.

     * @param _tokenId uint256 ID of the token to set user info for
     * @param _user address of the new user
     * @param _expires Unix timestamp when user info expires
     * @param _isBorrowed flag whether or not the NFT is borrowed
     */
    function setUser(uint256 _tokenId, address _user, uint64 _expires, bool _isBorrowed) external;

    /**
     * @notice Get the user address of an NFT.
     * @dev Reverts if user is not set.
     * @param _tokenId uint256 ID of the token to get the user address for
     * @return address user address for this NFT
     */
    function userOf(uint256 _tokenId) external view returns (address);

    /**
     * @notice Get the user expires of an NFT.
     * @param _tokenId uint256 ID of the token to get the user expires for
     * @return uint64 user expires for this NFT
     */
    function userExpires(uint256 _tokenId) external view returns (uint64);

    /**
     * @notice Get the user isBorrowed of an NFT.
     * @param _tokenId uint256 ID of the token to get the user isBorrowed for
     * @return uint64 user isBorrowed for this NFT
     */
    function userIsBorrowed(uint256 _tokenId) external view returns (bool);
}
```

> Every contract implementing the `IERC5501` interface is free to define the permissions of a `user`. However, user MUST NOT be considered an `owner`. They MUST NOT be able to execute transfers and approvals. Furthermore, `setUser` MUST be blocked from executing if `userIsBorrowed` returns `true` and `userExpires` is larger than or equal to `block.timestamp`.

Reviewer: ```suggestion

Every contract implementing the `IERC5501` interface is free to define the permissions of a `user`. However, `user` MUST NOT be considered an `owner`. They MUST NOT be able to execute transfers and approvals. Furthermore, `setUser` MUST be blocked from executing if `userIsBorrowed` returns `true` and `userExpires` is larger than or equal to `block.timestamp`.
```
Required: formatting.

The `UpdateUser` event MUST be emitted when a `user` is changed.   
The `setUser(uint256 _tokenId, address _user, uint64 _expires, bool _isBorrowed)` function SHOULD `revert` unless the `msg.sender` is the `owner` or an approved operator. It MUST revert if a token is borrowed and status has not expired yet. It MAY be `public` or `external`.   
The `userOf(uint256 _tokenId)` function SHOULD revert if `user` is not set or expired.   
The `userExpires(uint256 _tokenId)` function returns a timestamp when user status expires.   
The `userIsBorrowed(uint256 _tokenId)` function returns whether NFT is borrowed or not.   
The `supportsInterface` function MUST return `true` when called with `0xf808ec37`.   
On every `transfer`, the `user` MUST be reset if the token is not borrowed. If the token is borrowed the `user` MUST stay the same.   

> **The Balance extension is OPTIONAL. This gives the options to query amount of tokens user has.**

Reviewer: ```suggestion
**The Balance extension is OPTIONAL. This gives the option to query the number of tokens a `user` has.**

```
Required: formatting.
Required: grammar.

```solidity
/**
 * @title IERC5501Balance

>  * @dev See https://eips.ethereum.org/EIPS/eip-5501

Reviewer: ```suggestion
```
Required: no external links.

 * Extension for ERC5501 which adds userBalanceOf to query how many tokens address is userOf.
 * @notice the EIP-165 identifier for this interface is 0x0cb22289.
 */
interface IERC5501Balance /* is IERC5501 */{
    /**
     * @notice Count of all NFTs assigned to a user.
     * @dev Reverts if user is zero address.
     * @param _user an address for which to query the balance
     * @return uint256 the number of NFTs the user has
     */
    function userBalanceOf(address _user) external view returns (uint256);
}
```

> The `userBalanceOf(address _user)` function SHOULD `revert` for zero address.

Reviewer: ```suggestion

The `userBalanceOf(address _user)` function SHOULD `revert` for zero address.
```
Required: formatting.

> **The Enumerable extension is OPTIONAL. This allows to iterate over user balance.**

Reviewer: ```suggestion
**The Enumerable extension is OPTIONAL. This allows to iterate over user balance.**

```
Required: formatting.

```solidity
/**
 * @title IERC5501Enumerable
 * @dev See https://eips.ethereum.org/EIPS/eip-5501
 * This extension for ERC5501 adds the option to iterate over user tokens.
 * @notice the EIP-165 identifier for this interface is 0x1d350ef8.
 */
interface IERC5501Enumerable /* is IERC5501Balance, IERC5501 */ {
    /**
     * @notice Enumerate NFTs assigned to a user.
     * @dev Reverts if user is zero address or _index >= userBalanceOf(_owner).
     * @param _user an address to iterate over its tokens
     * @return uint256 the token ID for given index assigned to _user
     */
    function tokenOfUserByIndex(address _user, uint256 _index) external view returns (uint256);
}
```

> The `tokenOfUserByIndex(address _user, uint256 _index)` function SHOULD `revert` for zero address and `throw` if the index is larger than or equal to `user` balance.

Reviewer: ```suggestion

The `tokenOfUserByIndex(address _user, uint256 _index)` function SHOULD `revert` for zero address and `throw` if the index is larger than or equal to `user` balance.
```

> **The Terminable extension is OPTIONAL. This allows to terminate the rent early if both parties agree.**

Reviewer: ```suggestion
**The Terminable extension is OPTIONAL. This allows terminating the rent early if both parties agree.**

```
Required: formatting.
Required: gramamar.

```solidity
/**
 * @title IERC5501Terminable

>  * @dev See https://eips.ethereum.org/EIPS/eip-5501

Reviewer: ```suggestion
```
Required: no external links.

    /**
     * @dev Emitted when borrow of token ID is terminated.
     */
    event TerminateBorrow(uint256 indexed _tokenId, address indexed _lender, address indexed _borrower, address _caller);

    /**
     * @notice Agree to terminate a borrowing.
     * @dev Lender must be ownerOf token ID. Borrower must be userOf token ID.
     * If lender and borrower are the same, set termination agreement for both at once.
     * @param _tokenId uint256 ID of the token to set termination info for
     */
    function setBorrowTermination(uint256 _tokenId) external;

    /**
     * @notice Get if it is possible to terminate a borrow agreement.
     * @param _tokenId uint256 ID of the token to get termination info for
     * @return bool, bool first indicates lender agrees, second indicates borrower agrees
     */
    function getBorrowTermination(uint256 _tokenId) external view returns (bool, bool);

    /**
     * @notice Terminate a borrow if both parties agreed.
     * @dev Both parties must have agreed, otherwise revert.
     * @param _tokenId uint256 ID of the token to terminate borrow of
     */
    function terminateBorrow(uint256 _tokenId) external;
}
```

> The `AgreeToTerminateBorrow` event MUST be emitted when either the lender or borrower agrees to terminate the rent.   

Reviewer: ```suggestion

The `AgreeToTerminateBorrow` event MUST be emitted when either the lender or borrower agrees to terminate the rent.
```
Required: formatting.
